### PR TITLE
Add Undo and Redo Command

### LINKED ISSUE #83: Undo Command

### LINKED ISSUE #84: Redo Command

### PR DESCRIPTION
**Pull request by RadieonAjax**
#### **Summary**
This PR introduces the ability to undo and redo operations in the address book, ensuring users can easily revert or reapply changes made to their data.

#### **Key Changes:**
1. **Undo/Redo Functionality**:
   - Implemented `undoAddressBook()` to revert the address book to its previous state.
   - Implemented `redoAddressBook()` to restore the state after an undo.
   - Added safeguards to ensure these operations are only possible when there is a valid history to revert or reapply.

2. **State Management**:
   - Introduced `versionedAddressBook` to track the history of changes and manage undo/redo operations.
   - Integrated `canUndoAddressBook()` and `canRedoAddressBook()` to validate the feasibility of undo/redo actions.
   - Added `saveAddressBook()` to store the current state for future reference.

3. **Filter List Update**:
   - After every undo/redo action, the filtered list of persons is updated to reflect the latest state.

4. **Logging Enhancements**:
   - Added logging to track undo/redo actions, making it easier to debug and audit operations.

#### **Testing and Coverage**:
- Added test cases for `undoAddressBook()` and `redoAddressBook()` in the model.
- Verified that the filtered list properly updates after each undo/redo action.
- Ensured edge cases, such as undo/redo when there is no available history, are handled gracefully.

#### **Related Issues**:
- Closes #83 (Undo functionality)
- Closes #84 (Redo functionality)

### CODE REVIEW

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 1]
```diff
@@ -79,6 +82,52 @@ public void execute_invalidIndexFilteredList_throwsCommandException() {
...
+        showPersonAtIndex(model, INDEX_SECOND_PERSON);
+        Person personToDelete = model.getFilteredPersonList().get(INDEX_FIRST_PERSON.getZeroBased());
+        expectedModel.deletePerson(personToDelete);
+        expectedModel.saveAddressBook();
+        deleteCommand.execute(model);
+        expectedModel.undoAddressBook();
+        assertCommandSuccess(new UndoCommand(), model, UndoCommand.MESSAGE_UNDO_SUCCESS, expectedModel);
+        assertNotEquals(personToDelete, model.getFilteredPersonList().get(INDEX_FIRST_PERSON.getZeroBased()));
+        expectedModel.redoAddressBook();
+        assertCommandSuccess(new RedoCommand(), model, RedoCommand.MESSAGE_REDO_SUCCESS, expectedModel);
```
Comment (isaactodo) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:128]: Can add some blank lines to make it more readable, same with some of the other test cases

**Review by isaactodo — APPROVED** [review]
LGTM! Just the test case code a bit hard to read

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java (hunk 1 above)
Comment (RadieonAjax) [reply src/test/java/seedu/address/logic/commands/DeleteCommandTest.java]: Right, I've added blank lines in between to make it easier to see!
